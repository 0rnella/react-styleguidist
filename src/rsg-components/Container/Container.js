import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prop-types
function Render({ wrappers, props, children }) {
	const [wrapper, ...rest] = wrappers;

	if (rest.length > 0) {
		children = <Render wrappers={rest}>{children}</Render>;
	}

	const Wrapper = wrapper.render;
	return <Wrapper {...props}>{children}</Wrapper>;
}

export default function Container({ name, children, props = {} }, { slots }) {
	const fills = slots[name];
	if (!fills) {
		throw new Error(`Slot "${name}" not found, available slots: ${Object.keys(slots).join(', ')}`);
	}

	if (fills.length > 0) {
		return (
			<Render wrappers={fills} props={props}>
				{children}
			</Render>
		);
	}

	return children;
}

Container.propTypes = {
	name: PropTypes.string.isRequired,
	props: PropTypes.object,
	children: PropTypes.node,
};
Container.contextTypes = {
	slots: PropTypes.object.isRequired,
};
