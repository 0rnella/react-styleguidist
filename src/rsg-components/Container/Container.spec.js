import React from 'react';
import Container from './Container';

/* eslint-disable react/prop-types */

const children = <span>coffee</span>;
const options = {
	context: {
		slots: {
			slot: [
				{
					id: 'one',
					render: ({ children, ...rest }) => (
						<div {...rest} name="one">
							{children}
						</div>
					),
				},
				{
					id: 'two',
					render: ({ children, ...rest }) => (
						<div {...rest} name="two">
							{children}
						</div>
					),
				},
			],
			slot2: [],
		},
	},
};

it('should renderer slots and pass props and children', () => {
	const actual = render(
		<Container name="slot" props={{ title: 'Pizza' }}>
			{children}
		</Container>,
		options
	);

	expect(actual).toMatchSnapshot();
});

it('should return children if there are no slots', () => {
	const actual = render(<Container name="slot2">{children}</Container>, options);

	expect(actual).toMatchSnapshot();
});
