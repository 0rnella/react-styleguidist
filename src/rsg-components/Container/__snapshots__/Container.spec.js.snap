// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renderer slots and pass props and children 1`] = `
<div
  name="one"
  title="Pizza"
>
  <div
    name="two"
  >
    <span>
      coffee
    </span>
  </div>
</div>
`;

exports[`should return children if there are no slots 1`] = `
<span>
  coffee
</span>
`;
